File: bindings/r/R/strict_json.R
.strling_skip_whitespace <- function(text, index) {
  length <- nchar(text, type = "chars")
  while (index <= length && grepl(substr(text, index, index), " \t\r\n", fixed = TRUE)) index <- index + 1L
  index
}

.strling_scan_string <- function(text, index) {
  if (substr(text, index, index) != '"') stop("expected JSON string", call. = FALSE)
  start <- index
  index <- index + 1L
  length <- nchar(text, type = "chars")
  while (index <= length) {
    char <- substr(text, index, index)
    value <- utf8ToInt(char)
    index <- index + 1L
    if (char == '"') return(list(index = index, literal = substr(text, start, index - 1L)))
    if (length(value) == 1L && value < 32L) stop("unescaped control character in JSON string", call. = FALSE)
    if (char == "\\") {
      escape <- substr(text, index, index)
      index <- index + 1L
      if (escape == "u") {
        hex <- substr(text, index, index + 3L)
        if (!grepl("^[0-9a-fA-F]{4}$", hex)) stop("invalid Unicode escape", call. = FALSE)
        index <- index + 4L
      } else if (!grepl(escape, '"\\/bfnrt', fixed = TRUE)) {
        stop("invalid JSON escape", call. = FALSE)
      }
    }
  }
  stop("unterminated JSON string", call. = FALSE)
}

.strling_scan_value <- NULL

.strling_scan_object <- function(text, index) {
  index <- .strling_skip_whitespace(text, index + 1L)
  if (substr(text, index, index) == "}") return(index + 1L)
  keys <- new.env(hash = TRUE, parent = emptyenv())
  repeat {
    scanned <- .strling_scan_string(text, .strling_skip_whitespace(text, index))
    index <- scanned$index
    key <- jsonlite::fromJSON(scanned$literal, simplifyVector = FALSE)
    identity <- paste0("key:", key)
    if (exists(identity, envir = keys, inherits = FALSE)) stop(paste("duplicate property", scanned$literal), call. = FALSE)
    assign(identity, TRUE, envir = keys)
    index <- .strling_skip_whitespace(text, index)
    if (substr(text, index, index) != ":") stop("expected JSON colon", call. = FALSE)
    index <- .strling_scan_value(text, index + 1L)
    index <- .strling_skip_whitespace(text, index)
    delimiter <- substr(text, index, index)
    if (delimiter == "}") return(index + 1L)
    if (delimiter != ",") stop("expected JSON object delimiter", call. = FALSE)
    index <- index + 1L
  }
}

.strling_scan_array <- function(text, index) {
  index <- .strling_skip_whitespace(text, index + 1L)
  if (substr(text, index, index) == "]") return(index + 1L)
  repeat {
    index <- .strling_scan_value(text, index)
    index <- .strling_skip_whitespace(text, index)
    delimiter <- substr(text, index, index)
    if (delimiter == "]") return(index + 1L)
    if (delimiter != ",") stop("expected JSON array delimiter", call. = FALSE)
    index <- index + 1L
  }
}

.strling_scan_value <- function(text, index) {
  index <- .strling_skip_whitespace(text, index)
  char <- substr(text, index, index)
  if (char == "{") return(.strling_scan_object(text, index))
  if (char == "[") return(.strling_scan_array(text, index))
  if (char == '"') return(.strling_scan_string(text, index)$index)
  length <- nchar(text, type = "chars")
  finish <- index
  while (finish <= length && !grepl(substr(text, finish, finish), " \t\r\n,]}", fixed = TRUE)) finish <- finish + 1L
  if (finish == index) stop("invalid JSON token", call. = FALSE)
  finish
}

.strling_encode_json <- function(value) {
  jsonlite::toJSON(value, auto_unbox = TRUE, null = "null", na = "null", digits = NA, pretty = FALSE)
}

.strling_decode_object <- function(raw) {
  text <- rawToChar(raw)
  finish <- .strling_scan_value(text, 1L)
  if (.strling_skip_whitespace(text, finish) != nchar(text, type = "chars") + 1L) stop("trailing JSON content", call. = FALSE)
  value <- jsonlite::fromJSON(text, simplifyVector = FALSE)
  if (!is.list(value) || is.null(names(value))) stop("interop response must be an object", call. = FALSE)
  value
}

File: bindings/php/src/StrictJson.php
<?php

declare(strict_types=1);

namespace STRling;

/** Strict JSON codec with duplicate-property rejection. */
final class StrictJson
{
    /** @param mixed $value */
    public static function encode($value): string
    {
        return json_encode($value, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    }

    /** @return array<string, mixed> */
    public static function decodeObject(string $json): array
    {
        $offset = 0;
        self::scanValue($json, $offset);
        self::skipWhitespace($json, $offset);
        if ($offset !== strlen($json)) {
            throw new \JsonException('trailing JSON content');
        }
        $value = json_decode($json, true, 512, JSON_THROW_ON_ERROR | JSON_BIGINT_AS_STRING);
        if (!is_array($value) || array_is_list($value)) {
            throw new \JsonException('interop response must be an object');
        }
        return $value;
    }

    private static function scanValue(string $json, int &$offset): void
    {
        self::skipWhitespace($json, $offset);
        $char = $json[$offset] ?? '';
        if ($char === '{') {
            self::scanObject($json, $offset);
            return;
        }
        if ($char === '[') {
            self::scanArray($json, $offset);
            return;
        }
        if ($char === '"') {
            self::scanString($json, $offset);
            return;
        }
        $remaining = substr($json, $offset);
        if (preg_match('/\A(?:true|false|null)/', $remaining, $match) === 1) {
            $offset += strlen($match[0]);
            return;
        }
        if (preg_match('/\A-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/', $remaining, $match) === 1) {
            $offset += strlen($match[0]);
            return;
        }
        throw new \JsonException(sprintf('invalid JSON token at byte %d', $offset));
    }

    private static function scanObject(string $json, int &$offset): void
    {
        ++$offset;
        self::skipWhitespace($json, $offset);
        if (($json[$offset] ?? '') === '}') {
            ++$offset;
            return;
        }
        $keys = [];
        while (true) {
            self::skipWhitespace($json, $offset);
            $literal = self::scanString($json, $offset);
            $key = json_decode($literal, true, 2, JSON_THROW_ON_ERROR);
            $identity = "\0" . $key;
            if (isset($keys[$identity])) {
                throw new \JsonException(sprintf('duplicate property %s', $literal));
            }
            $keys[$identity] = true;
            self::skipWhitespace($json, $offset);
            if (($json[$offset] ?? '') !== ':') {
                throw new \JsonException(sprintf('expected colon at byte %d', $offset));
            }
            ++$offset;
            self::scanValue($json, $offset);
            self::skipWhitespace($json, $offset);
            $delimiter = $json[$offset] ?? '';
            if ($delimiter === '}') {
                ++$offset;
                return;
            }
            if ($delimiter !== ',') {
                throw new \JsonException(sprintf('expected object delimiter at byte %d', $offset));
            }
            ++$offset;
        }
    }

    private static function scanArray(string $json, int &$offset): void
    {
        ++$offset;
        self::skipWhitespace($json, $offset);
        if (($json[$offset] ?? '') === ']') {
            ++$offset;
            return;
        }
        while (true) {
            self::scanValue($json, $offset);
            self::skipWhitespace($json, $offset);
            $delimiter = $json[$offset] ?? '';
            if ($delimiter === ']') {
                ++$offset;
                return;
            }
            if ($delimiter !== ',') {
                throw new \JsonException(sprintf('expected array delimiter at byte %d', $offset));
            }
            ++$offset;
        }
    }

    private static function scanString(string $json, int &$offset): string
    {
        $start = $offset;
        if (($json[$offset] ?? '') !== '"') {
            throw new \JsonException(sprintf('expected string at byte %d', $offset));
        }
        ++$offset;
        $length = strlen($json);
        while ($offset < $length) {
            $char = $json[$offset++];
            if ($char === '"') {
                return substr($json, $start, $offset - $start);
            }
            if (ord($char) < 0x20) {
                throw new \JsonException('unescaped control character in string');
            }
            if ($char !== '\\') {
                continue;
            }
            $escape = $json[$offset++] ?? '';
            if ($escape === 'u') {
                $hex = substr($json, $offset, 4);
                if (strlen($hex) !== 4 || preg_match('/\A[0-9a-fA-F]{4}\z/', $hex) !== 1) {
                    throw new \JsonException('invalid Unicode escape');
                }
                $offset += 4;
            } elseif (!str_contains('"\\/bfnrt', $escape)) {
                throw new \JsonException('invalid string escape');
            }
        }
        throw new \JsonException('unterminated string');
    }

    private static function skipWhitespace(string $json, int &$offset): void
    {
        $length = strlen($json);
        while ($offset < $length && str_contains(" \t\r\n", $json[$offset])) {
            ++$offset;
        }
    }
}
